src/atom_table.rs

```diff
@@ -194,41 +253,53 @@ impl Atom {
         let c1 = it.next();
         let c2 = it.next();
 
-        if c2.is_none() { c1 } else { None }
+        if c2.is_none() {
+            c1
+        } else {
+            None
+        }
     }
 
-    #[inline]
-    pub fn chars(&self) -> str::Chars {
-        self.as_str().chars()
+    #[inline(always)]
+    pub fn as_str_with_table<'at>(&self, atom_table: &'at AtomTable) -> &'at str {
+        if let Some(ptr) = self.as_ptr_with_table(atom_table) {
+            let header = unsafe { ptr::read::<AtomHeader>(ptr as *const u8 as *const AtomHeader) };
+            let len = header.len() as usize;
+            let buf = (unsafe { (ptr as *const u8).offset(mem::size_of::<AtomHeader>() as isize) })
+                as *mut u8;
+
+            unsafe { str::from_utf8_unchecked(slice::from_raw_parts(buf, len)) }
+        } else {
+            &STRINGS[(self.index >> 3) as usize]
+        }
     }
 
+    #[track_caller]
     #[inline]
-    pub fn as_str(&self) -> &str {
-        unsafe {
-            let ptr = self.as_ptr();
-
-            if ptr.is_null() {
-                return STRINGS[(self.index >> 3) as usize];
-            }
-
-            let header = ptr::read::<AtomHeader>(ptr as *const _);
-            let len = header.len() as usize;
-            let buf = (ptr as usize + mem::size_of::<AtomHeader>()) as *mut u8;
-
-            str::from_utf8_unchecked(slice::from_raw_parts(buf, len))
+    pub fn as_str(&self) -> AtomString<'static> {
+        if self.is_static() {
+            AtomString::Static(STRINGS[(self.index >> 3) as usize])
+        } else {
+            let guard = owned_atom_table_read_guard()
+                .expect("We should only have an Atom while there is an AtomTable");
+            AtomString::Dynamic(OwnedRwLockReadGuard::map(guard, |atom_table| {
+                self.as_str_with_table(atom_table)
+            }))
         }
     }
 
-    pub fn defrock_brackets(&self, atom_tbl: &mut AtomTable) -> Self {
+    pub fn defrock_brackets(&self, atom_tbl: &Arc<RwLock<AtomTable>>) -> Self {
         let s = self.as_str();
 
-        let s = if s.starts_with('(') && s.ends_with(')') {
+        let sub_str = if s.starts_with('(') && s.ends_with(')') {
             &s['('.len_utf8()..s.len() - ')'.len_utf8()]
         } else {
             return *self;
         };
 
-        atom_tbl.build_with(s)
+        let val = sub_str.to_string();
+        drop(s); // wee need to drop s as it holds a read lock on the AtomTable and build_with may need to acquire a write lock
+        AtomTable::build_with(&atom_tbl, &val)
```

Comment:
I think this was a bug before, in the case that build_with needed to grow the RawBlock. the &str passed to build_with would point to deallocated memory.